### PR TITLE
[ENH] For classification: Adding InceptionTime, CNN and ResNet along with other classifiers and removing LSTM-FCN

### LINKED ISSUE #106: [ENH] Port in InceptionTime for classification and regression
InceptionTime is a deep learning algorithm for classification and regression. We have versions here 

https://github.com/time-series-machine-learning/tsml-eval/tree/main/tsml_eval/sktime_estimators

that we would like ported over. Ali has agreed to do it, so just raising an issue as a place holder and assigned myself, it has already been claimed, soz!

edit: Also going to include an option to run with the improvements they proposed here

https://maxime-devanne.com/pages/HCCF-4-tsc/

### PR DESCRIPTION
**Pull request by hadifawaz1999**
#### Reference Issues/PRs 

Added InceptionTime for classification as mentioned in #106 with the option to use [custom filters](https://maxime-devanne.com/pages/HCCF-4-tsc/).

Fixes #106 

#### What does this implement/fix? Explain your changes.

 I made the models CNN, ResNet and InceptionTime parametrized so the users can change in the architecture how they like, default parameters are taken from [here](https://github.com/hfawaz/dl-4-tsc). Some comments about some code lines in CNN are added, noticed some questions about them.

I started by adding missing classifiers from the original [review](https://github.com/hfawaz/dl-4-tsc) such as Encoder. More classifiers will be added on the same branch.

I removed LSTM-FCN classifier, given the issues about monitoring on the test set. The model as well has a feature of applying LSTM on 1 time stamp.
For more information, the following issues should be addressed:

1. [monitoring the val loss](https://github.com/titu1994/LSTM-FCN/issues/7#issuecomment-423917477)
2. [lstm on 1 time stamp](https://github.com/titu1994/LSTM-FCN/issues/4#issue-323439220)

#### Does your contribution introduce a new dependency? If yes, which one?

yes [tensorflow_addons](https://github.com/tensorflow/addons) for the Encoder classifier.

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/.all-contributorsrc).
- [ ] Optionally, I've updated sktime's [CODEOWNERS](https://github.com/sktime/sktime/blob/main/CODEOWNERS) to receive notifications about future changes to these files.
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG] indicating whether the PR topic is related to enhancement, maintenance, documentation, or bug.

##### For new estimators
- [ ] I've added the estimator to the online documentation.
- [ ] I've updated the existing example notebooks or provided a new one to showcase how my estimator works.

### CODE REVIEW

**Review by TonyBagnall — CHANGES_REQUESTED** [review]
this is great, but needs a bit of work to conform. Firstly, InceptionTime should inherit from BaseClassifier and implement    _fit/_predict not fit/predict. Secondly, for scikit learn compliance, you should not change the value stored in variables passes as arguments to the constructor, you should store in a new private variable, so rather than this in the constructor

     if isinstance(kernel_sizes, list):
            self.kernel_sizes = kernel_sizes
        else:
            self.kernel_sizes = [kernel_sizes] * n_layers

you would do this in the constructor
self.kernel_sizes = kernel_sizes

then in fit you do this
     if isinstance(self.kernel_sizes, list):
            self._kernel_sizes = self.kernel_sizes
        else:
            self._kernel_sizes = [self.kernel_sizes] * n_layers

its for things like CV and definitely a scikit thing not ours :) At least as I understand it, someone else will tell me if thats incorrect.

Only fields in the child class need to be listed in the doc string/

File: sktime/classification/deep_learning/cnn.py [hunk 1]
```diff
@@ -20,34 +20,48 @@ class CNNClassifier(BaseDeepClassifier):
 
     Parameters
     ----------
-    should inherited fields be listed here?
```
[the remark this answers is not in the document]
Comment (hadifawaz1999) [reply sktime/classification/deep_learning/cnn.py:23]: Thank you for you comments, I did the necessary changes:

1. InceptionTime now inherits from BaseClassifier with _fit _predict and _predict_proba as functions instead of fit predict and predict_proba
2. Took care of the change in argument variables changing and set them as private variables before assignment

**Review by TonyBagnall — CHANGES_REQUESTED** [review]
great, a few more points, some are improvements to existing code, but since we are here :)
1. a couple of formatting issues to sort out, failing code quality.  
2. the next ask is correctness testing. This is put in place to make sure the algorithm doesn't get changed without people noticing. Look in classification/tests/ for the model for it, ask on slack if its not clear
3. The access format in the existing deep learning is different to the others, so may as well change here. If you look in, for example, interval based, the classifiers are private and accessed through the init method. Good to have the same pattern in all
4. why are there separate files lstmfcn and lstfcm_layers in networks? coould the layers go in the same file?

**Comment by hadifawaz1999** [comment]
Okay I will go through all that. For now I took care of some code quality errors, not sure why locally they did not appear before.
For lstm fcn, I think it does not matter, because I actually deleted it in my branch if am not mistaken and added in the comment of the PR why I removed it. Am not sure why it was implemented as lstmfcn and lstmfcn_layer independently though.

**Review by MatthewMiddlehurst — CHANGES_REQUESTED** [review]
I think the InceptionTime network is missing some dependency tags i.e. `_tags = {"python_dependencies": "tensorflow"}` from CNN.

image

File: sktime/classification/deep_learning/encoder.py [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+            `MyClass(**params)` or `MyClass(**params[i])` creates a valid test instance.
+            `create_test_instance` uses the first (or only) dictionary in `params`.
+        """
+        from sktime.utils.validation._dependencies import _check_soft_dependencies
+
+        param1 = {
+            "n_epochs": 10,
+            "batch_size": 4,
+            "avg_pool_size": 4,
+        }
```
Comment (MatthewMiddlehurst) [inline sktime/classification/deep_learning/encoder.py:252]: "avg_pool_size" is not a parameter which is causing one of the test failures.

**Comment by MatthewMiddlehurst** [comment]
Thanks for contributing! There may be a couple more comments/requests, but best to get the current CI failures out of the way.

Feel free to ask if there are any questions.

File: sktime/classification/deep_learning/cnn.py [hunk 3]
```diff
@@ -20,34 +20,51 @@ class CNNClassifier(BaseDeepClassifier):
...
+    dilation_rate   : int or list of int, default = 1,
+        the dilation rate of the convolution layers, if not a list,
+        the same dilation rate is used all over the network
+    use_bias        : bool or list of bool, default = True,
+        condition on wether or not to use bias values for convolution layers,
+        if not a list, the same condition is used for all layers
+    random_state    : int, default = 0
+        seed to any needed random actions
+    nb_epochs       : int, default = 2000
         the number of epochs to train the model
```
[the remark this answers is not in the document]
Comment (hadifawaz1999) [reply sktime/classification/deep_learning/cnn.py:55]: Actually for this parameter, its easier if my code adapts the notebooks, so which one would you want me to use? I will use it through all my branch.

[the remark this answers is not in the document]
Comment (TonyBagnall) [reply sktime/classification/deep_learning/cnn.py:55]: I would prefer n_epochs I think, since its that already

[the remark this answers is not in the document]
Comment (hadifawaz1999) [reply sktime/classification/deep_learning/cnn.py:55]: okay i will make the change, thanks

File: sktime/classification/deep_learning/inception_time.py [hunk 4]
```diff
@@ -0,0 +1,673 @@
...
+from sktime.networks.inception import InceptionNetwork
+from sktime.utils.validation._dependencies import _check_dl_dependencies
+
+_check_dl_dependencies(severity="warning")
+
+
+class InceptionTimeClassifier(BaseClassifier):
+    """InceptionTime ensemble classifier.
+
+    Ensemble of IndividualInceptionTimeClassifiers, as desribed in [1].
```
Comment (TonyBagnall) [inline sktime/classification/deep_learning/inception_time.py:21]: described

File: sktime/classification/deep_learning/inception_time.py [hunk 5]
```diff
@@ -0,0 +1,673 @@
...
+            connections all over Inception
+        use_bottleneck      : bool, default = True,
+            confition wether or not to use bottlesnecks
+            all over Inception
+        bottleneck_size     : int, default = 32,
+            the bottleneck size in case use_bottleneck = True
+        use_custom_filters  : bool, default = True,
+            condition on wether or not to use custom
+            filters in the first inception module
+
```
Comment (TonyBagnall) [inline sktime/classification/deep_learning/inception_time.py:84]: no blank line

File: sktime/classification/deep_learning/inception_time.py [hunk 6]
```diff
@@ -0,0 +1,673 @@
...
+    ..[2] Ismail-Fawaz et al. Deep Learning For Time Series
+    Classification Using New
+    Hand-Crafted Convolution Filters, 2022 IEEE International
+    Conference on Big Data.
+
+    Adapted from the implementation from Fawaz et. al
+    https://github.com/hfawaz/InceptionTime/blob/master/classifiers/inception.py
+
+    and Ismail-Fawaz et al.
+    https://github.com/MSD-IRIMAS/CF-4-TSC
```
Comment (TonyBagnall) [inline sktime/classification/deep_learning/inception_time.py:121]: an Example would be good

File: sktime/classification/deep_learning/inception_time.py [hunk 7]
```diff
@@ -0,0 +1,673 @@
...
+        self.random_state = random_state
+        self.verbose = verbose
+        self.use_mini_batch_size = use_mini_batch_size
+        self.loss = loss
+        self.metrics = metrics
+        self.optimizer = optimizer
+
+        self.classifers_ = []
+
+        # super(InceptionTimeClassifier, self).__init__()
```
Comment (TonyBagnall) [inline sktime/classification/deep_learning/inception_time.py:188]: how come this is commented out?

File: sktime/classification/deep_learning/inception_time.py [hunk 8]
```diff
@@ -0,0 +1,673 @@
...
+                loss=self.loss,
+                metrics=self.metrics,
+                optimizer=self.optimizer,
+                random_state=rng.randint(0, np.iinfo(np.int32).max),
+                verbose=self.verbose,
+            )
+            cls.fit(X, y)
+            self.classifers_.append(cls)
+
+            self.classes_ = cls.classes_
```
Comment (TonyBagnall) [inline sktime/classification/deep_learning/inception_time.py:238]: you dont need to set this inside the loop, in fact its set in base class fit()

File: sktime/classification/deep_learning/inception_time.py [hunk 9]
```diff
@@ -0,0 +1,673 @@
...
+
+        return self
+
+    def _predict(self, X) -> np.ndarray:
+        """Predict the labels of the test set using InceptionTime.
+
+        Arguments:
+        ---------
+
+        X : np.ndarray of shape = (n_instances (n), n_dimensions (d), series_length (m))
```
Comment (TonyBagnall) [inline sktime/classification/deep_learning/inception_time.py:249]: changing to call them channels

File: sktime/classification/deep_learning/inception_time.py [hunk 10]
```diff
@@ -0,0 +1,673 @@
...
+
+    def build_model(self, input_shape, n_classes, **kwargs):
+        """
+        Construct a compiled, un-trained, keras model that is ready for training.
+
+        Parameters
+        ----------
+        input_shape : tuple
+            The shape of the data fed into the input layer
+        nb_classes: int
```
Comment (TonyBagnall) [inline sktime/classification/deep_learning/inception_time.py:501]: n_classes

File: sktime/classification/deep_learning/resnet.py [hunk 11]
```diff
@@ -21,27 +21,54 @@ class ResNetClassifier(BaseDeepClassifier):
...
+        activation                  : str or list of str, default = 'relu',
+            keras activation used in the convolution layers
+            in one residual block, if not
+            a list, the same kernel size is used in all convolution layers
+        use_bias                    : bool or list of bool, default = True,
+            condition on wether or not to use bias values in
+            the convolution layers in one residual block, if not
+            a list, the same kernel size is used in all convolution layers
+
+        nb_epochs                   : int, default = 1500
```
Comment (TonyBagnall) [inline sktime/classification/deep_learning/resnet.py:55]: we are going with n_epochs I think

**Review by TonyBagnall — CHANGES_REQUESTED** [review]
so minor things, looking good

**Comment by hadifawaz1999** [comment]
Thanks @TonyBagnall , I fixed all the problems you mentioned, but for the example I had an issue with code quality (line too long when importing inceptionTime classifier) so I had to make it on two lines. Let me know if everything is good.

**Comment by hadifawaz1999** [comment]
If i do git pull main while in my branch locally wont it remove my changes ?

**Comment by MatthewMiddlehurst** [comment]
It shouldn't, no. There is a single conflict that will have to be resolved but that can be done by deleting `lstmfcn.py` again.

image

**Comment by MatthewMiddlehurst** [comment]
Most of the changes will be merged automatically, if there is a local change from you and a change on main on the same file it should prompt you to resolve it and commit again before pushing.

**Comment by TonyBagnall** [comment]
File /opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/sktime/classification/base.py:192, in BaseClassifier.fit(self, X, y)
    187         raise AttributeError(
    188             "self.n_jobs must be set if capability:multithreading is True"
    189         )
    191 # pass coerced and checked data to inner _fit
--> 192 self._fit(X,y)
    193 self.fit_time_ = int(round(time.time() * 1000)) - start
    195 # this should happen last

I dont think you want to set multithreaded to be true, it will run automatically on the GPU, n_jobs is for CPU threading

File: sktime/classification/deep_learning/inception_time.py (hunk 6 above)
Comment (hadifawaz1999) [reply sktime/classification/deep_learning/inception_time.py:120]: an Example was added with fit and predict functions used

File: sktime/classification/deep_learning/inception_time.py [hunk 12]
```diff
@@ -0,0 +1,682 @@
...
+
+        for cls in self.classifers_:
+            probs += cls._predict_proba(X)
+
+        probs = probs / self.n_classifiers
+
+        return probs
+
+    def get_test_params(cls, parameter_set="default"):
+        """Return testing parameter settings for the estimator.
```
Comment (MatthewMiddlehurst) [inline sktime/classification/deep_learning/inception_time.py:297]: We are missing `@classmethod` at the top of this method.

File: 02a_classification_multivariate_cnn.ipynb [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+   "metadata": {
+    "pycharm": {
+     "is_executing": true
+    }
+   },
+   "outputs": [],
+   "source": [
+    "param_grid = {\"kernel_size\": [7, 9], \"avg_pool_size\": [3, 5]}\n",
+    "grid = GridSearchCV(network, param_grid=param_grid, cv=5)\n",
+    "grid.fit(X_train, y_train)\n",
```
Comment (MatthewMiddlehurst) [inline 02a_classification_multivariate_cnn.ipynb:105]: CNN has `kernel_sizes` but call is `kernel_size`.

**Review by MatthewMiddlehurst — COMMENTED** [review]
Think these are the test failures.

File: 02a_classification_multivariate_cnn.ipynb (hunk 13 above)
Comment (hadifawaz1999) [reply 02a_classification_multivariate_cnn.ipynb:105]: oh yeah most of the architectures now use kernel_sizes, I can change it back to kernel_size if its best for everyone, I made it with 's' because its more than one, but I think its okay without the 's' :) what do you think @MatthewMiddlehurst @TonyBagnall ?

File: sktime/classification/deep_learning/inception_time.py (hunk 12 above)
Comment (hadifawaz1999) [reply sktime/classification/deep_learning/inception_time.py]: Yeah I pushed now the fixed code for that in both InceptionTimeClassifier and IndividualInceptionClassifier methods

File: 02a_classification_multivariate_cnn.ipynb (hunk 13 above)
Comment (TonyBagnall) [reply 02a_classification_multivariate_cnn.ipynb:105]: I think best use the same for all, I dont mind which, if you choose sizes just change the notebook

Comment (hadifawaz1999) [reply 02a_classification_multivariate_cnn.ipynb:105]: I changed everything back to kernel_size

**Review by TonyBagnall — APPROVED** [review]

**Review by MatthewMiddlehurst — COMMENTED** [review]
Test parameter in CNN needs to be changed to `n_layers` from `n_conv_layers` I believe.

**Comment by MatthewMiddlehurst** [comment]
The classifier `get_test_params` had it has a parameter, but the class itself does not. This is breaking the tests for CNNClassifier

**Comment by hadifawaz1999** [comment]
@MatthewMiddlehurst I added the ```n_conv_layers``` parameter to the CNN classifier set by default to None to avoid test fails.

**Comment by MatthewMiddlehurst** [comment]
Hello, I think the changes in https://github.com/hadifawaz1999/scikit-time/pull/1 should fix the tests. There is a bit of a mixup beween `n_layers` and `n_conv_layers`. 

Also, not so relevant with those changes, but `n_conv_layers=None` as an argument must have an accompanying `self.n_conv_layers=n_conv_layers` in the init if it were to be included. Tests dont like arguments which go nowhere apparently 🙂.

**Comment by hadifawaz1999** [comment]
Hello @MatthewMiddlehurst , I made the changes as you said, made ```n_conv_layers``` a private variable in both the CNNNetwork and CNNClassifier classes, lets hope it works now

**Comment by hadifawaz1999** [comment]
Oh yeah that's my bad i will remove it, it was something i was testing for environments problems. Will push the branch with removing it when the tests are done.

**Comment by TonyBagnall** [comment]
ah this fail is known, we need to put InceptionTime on the excluded list. Go to tests/_config.py and add it here 
image

**Comment by hadifawaz1999** [comment]
on all classifiers the pickle problem exists @MatthewMiddlehurst @TonyBagnall , should we ignore it for all classifiers ?

**Review by MatthewMiddlehurst — APPROVED** [review]
I think this should be good (hopefully). A bit more of a pain than I thought it would be, but probably the hardest step 🙂. 

Another time it would be good to chip away at these tests individually and figure out A. what is breaking them and B. if it's something we actually can/want to fix. First 3 probably more difficult, last 2 I think easier.

- "test_fit_idempotent",
- "test_persistence_via_pickle"
- "test_save_estimators_to_file"
- "test_fit_does_not_overwrite_hyper_params"
- "test_methods_have_no_side_effects"

**Comment by hadifawaz1999** [comment]
I think a first step would be to adapt the tests parameters to the new ones. like the n_conv_layers in cnn

**Comment by MatthewMiddlehurst** [comment]
We should rethink the testing parameters and what the parameters are called, yeah. Ideally we want them to be quick as possible, and we could possibly use the new parameters to help with that. As you are probably aware at this point, the tests take an age to run and one of our goals is to reduce that by as much as possible 🙂. The issue is finding parameter values which cover as much as possible while remaining fast to run.

This issue we ran into in this PR was we renamed `n_conv_layers`, but kept that name in `get_test_parameters` i.e.

>         param1 = {
>             "n_epochs": 10,
>             "batch_size": 4,
>             "avg_pool_size": 4,
>             "n_conv_layers": 1,
>         }

So in the test suite, it would try to create a `CNNClassifier` and set `n_conv_layers` to 1, but would raise an error because `n_conv_layers` doesn't exist (as it was renamed).
